src/tools/compiletest/src/runtest.rs

```diff
@@ -3137,21 +3137,24 @@ impl<'test> TestCx<'test> {
                     self.fatal_proc_rec("test run succeeded!", &proc_res);
                 }
             }
+            if !self.props.error_patterns.is_empty() {
+                // "// error-pattern" comments
+                self.check_error_patterns(&proc_res.stderr, &proc_res);
```

Comment:
It's a bit weird that this is checked in two places - first in `should_run` and then in `!should_run` below.
I think, we should run `check_expected_errors` and `check_error_patterns` together under the same conditions.

Reply:
Also, could you try to remove the `!explicit && self.config.compare_mode.is_none()` condition as well?
I'm not sure what will happen, but perhaps we can check annotations in those modes as well?

Reply:
After fixing the first bug, i.e., no validation being performed when
compilation succeeds, some test cases started to fail because they expected
error patterns to be matched against program output. 

To address that issue, in the should_run case the error patterns are matched
against the program output, and otherwise the errors patterns are matched
against compiler output as before. This is why they are checked in two places.

Reply:
The `explicit` tests are those that use `--error-format` flag. They are
generally incompatible with error annotations, since there is only json error
parser. I also got two failures after removing this check, on `missing-type.rs`
and `unused_parens_remove_json_suggestion.rs`.

I am not familiar with compare mode, so I would rather not touch
`self.config.compare_mode.is_none()` condition either.

Reply:
>some test cases started to fail because they expected
error patterns to be matched against program output.

I see.
I'm pretty sure that was a misuse of the `// error-pattern` annotation.
We have a separate annotation for checking the runtime output - `// check-run-results`.

Reply:
Indeed, right now there are only handful of test in ui that use error-pattern
for matching executable results, so they could be presumably migrated to use
check-run-results.

At the same time, the changes are party motivated as help with #65506 which
moves run-fails tests to ui. For those tests such use of error-pattern is quite
typical.

Which is why I think we should allow for such use of error-pattern, at least
until all those tests are migrated if check-run-results is preferred variant.

Reply:
@tmiasko
Could you make an issue for migrating run-fail tests to `// check-run-results` and restoring the compile-time meaning of `// error-pattern` in them?